Detect support for protobuf array ref on server on frontend return call

hinshun opened the pull request:
Fixes #1327
Fixes #1321 
Regression introduced by: https://github.com/moby/buildkit/pull/1269

When using frontend built from regression commit and beyond, the frontend does not check for a server capability for ref arrays and sends ref arrays to the return API.

This PR, adds the apicap `CapProtoRefArray` and sends the appropriate ref/refarray to the return API depending on the apicap.

---

Reproduced with following steps:

Build the frontend from `master` of this repository:
```sh
$ DOCKER_BUILDKIT="1" docker build -t docker/dockerfile-upstream:regress -f ./frontend/dockerfile/cmd/dockerfile-frontend/Dockerfile .
```

Then building an image using `Docker 19.03` (with an older buildkitd):
```Dockerfile
# syntax=docker/dockerfile-upstream:regress
FROM alpine
```

Inspecting the image:
```sh
$ docker inspect -f "{{.RootFS.Layers}}" reproduction:regress
[]
```

With the changes in this PR:
```sh
$ docker inspect -f "{{.RootFS.Layers}}" reproduction:regress
[sha256:77cae8ab23bf486355d1b3191259705374f4a11d483b24964d2f729dd8c076a0]
```

tonistiigi commented on frontend/gateway/grpcclient/client.go:
```diff
@@ -105,22 +105,39 @@ func (c *grpcClient) Run(ctx context.Context, f client.BuildFunc) (retError erro
-						if err != nil {
-							retError = err
-							continue
+					if c.caps.Supports(opspb.CapProtoRefArray) == nil {
```
I think these are incompatible sets here. `c.caps` is gateway caps and `opspb.CapProtoRefArray` looks to be llb cap. Should be gateway cap.

hinshun commented:
We ended up agreeing to have 2 caps to not block PRS #1286 and #1290.
Updated the PR to include comment clarifying usage of `CapProtoRefArray` and moved capability to gateway cap package.
